pybossa/util.py

```diff
@@ -1368,26 +1368,43 @@ def extract_task_info_data(task, task_info_str):
         request_fields = request_fields.get(attribute, {})
     return json.dumps(request_fields)
 
-
-def get_user_saved_partial_tasks(sentinel, project_id, user_id):
+def get_user_saved_partial_tasks(sentinel, project_id, user_id, task_repo=None):
     """
-    Get the user saved task id list from Redis sorted set
+    Get the user saved task id list from Redis keys;
+    When "task_repo" is passed, it will do a cleanup of keys if tasks
+    have been deleted
     """
-    partial_answer_tasks = PARTIAL_ANSWER_TASKS.format(project_id=project_id,
-                                                       user_id=user_id)
-    now_timestamp = int(time.time())
-    task_ids = sentinel.master.zrangebyscore(partial_answer_tasks,
-                                             min=now_timestamp,
-                                             max=float('inf'),
-                                             withscores=True)
+    pattern = PARTIAL_ANSWER_KEY.format(project_id=project_id,
+                                        user_id=user_id,
+                                        task_id="*")
+    batch_size = 10000
+    keys = list(sentinel.slave.scan_iter(match=pattern, count=batch_size))
 
     result = dict()
-    for task_id_str, score in task_ids:
+    for k in keys:
         try:
-            task_id = int(task_id_str.decode('utf-8'))
-            ttl = int(score)
-            result[task_id] = ttl
+            k = k.decode('utf-8')
+            task_id = int(k.split(':')[-1])
+            ttl = sentinel.slave.ttl(k)
+            if task_repo:
+                task = task_repo.get_task(task_id)
```

Comment:
This loop has potential of making 10k sql requests to validate task. If we combine task_ids and validate tasks, 10k sql requests can be reduced to 1 sql request. Looks like there's scope of code refactoring here thereby reducing stress on db.

Reply:
Even with max 30 calls, there's potential of reducing sql queries by 30x. Hence, I recommend we update the logic to build list of tasks to be queried and issue single SQL call. Make sense?